### ISSUE TITLE
Proposal: improved semantics for `noreturn`-like types

### ISSUE THREAD

**Issue by mlugg** [ziglang/zig#15909]
This is in a sense a continuation of #3257. That issue laid some of the groundwork for the current status of `noreturn` and similar types, but there are a few things missing. Also, thanks to @ifreund for helping me figure out some of the details of this.

Lots of types can be constructed which are similar to `noreturn`. In type theory terms, they are "empty types": types for which no values exist. I'll call these types "noreturn-like". Here's an exhaustive list of types which should be considered noreturn-like (note the special cases around pointers mentioned at the end):
- `noreturn`
- `enum {}`
- `enum(T) {}` for any `T`
- `error {}`
- `union` / `extern union` where all fields are noreturn-like (including the case where the union has no fields)
- tagged unions where the tag type is noreturn-like or all fields are noreturn-like
- structs containing any noreturn-like field
- `E!T` where `E` and `T` are both noreturn-like (so, `error{}!T`)
- `[n]T` / `@Vector(n, T)` where `T` is noreturn-like and `n != 0`
- `*T` / `[*]T` (and any qualifiers) where `T` is noreturn-like
    - `[]T` has the value `&.{}` so is not noreturn-like, but rather zero-bit OPV
    - `[*c]T` has the value `null` so is not noreturn-like. In principle it's zero-bit OPV, but this is a footgun, so instead is disallowed.

Currently, most of these aren't "truly" considered noreturn-like. Such types should terminate semantic analysis when found (akin to `unreachable`). Additionally, `undefined` should not coerce to noreturn-like types, to avoid being able to actually "construct" them without terminating analysis.

We currently have the restriction that (at function scope) `var x: T = ...`, where `T` is noreturn-like, is a compile error. This, however, is inconsistent - `const x: T = ...` is allowed, and moreover, so is `var x = @as(T, ...)`. I propose that we remove this needless restriction. Allocs for noreturn-like types can be special-cased to emit no AIR instructions, and analysis will terminate before we attempt to store to one, as by that point we will have had an instruction whose result type is the child type.

Note, however, that global `var`s may not be of a noreturn-like type, nor may a value of a noreturn-like type be exported or marked extern. This is a compile error.

One last notable change in semantics here is that switching on a noreturn-like enum is no longer supported in the same way that it is in status quo. The idea currently is that you turn such an enum into the true `noreturn` type using an empty switch, `switch (e) {}`. This code will still be accepted under this proposal, but for a different reason: it's never analyzed! The moment `e` is created, semantic analysis terminates, since we reached a noreturn instruction.

**Comment by Vexu** [comment]
Duplicate/related #13807

**Comment by ifreund** [comment]
I think we need to reconsider the proposed semantics of pointers to noreturn types. As @jacobly0 pointed out, a pointer isn't required to point to valid memory unless it is dereferenced. As I see it, the logical extension of those semantics for pointers to noreturn types is that a pointer of type `*noreturn` may exist with arbitrary value. However, dereferencing a pointer of type `*noreturn` is a compile error as that would produce a value of type `noreturn` (which does not exist since noreturn is an empty type).

This also resolves an inconsistency in the current proposal when viewing a slice as a `struct { ptr: [*]noreturn, len: usize }`. By the struct rule and originally proposed pointer rule this should be a noreturn type. However, there is a valid possible value of this type, the empty slice.

Another issue with the current semantics that @jacobly0 brought up is that determining if a pointer is noreturn equivalent or not could cause dependency loops, e.g. due to needing to analyze struct field types in order to determine if a struct is noreturn equivalent.

To summarize for all pointer types:

- `*T`, `[*]T`, `[*c]T` are not empty types/"noreturn-like"
- Derefencing a pointer to noreturn-like type is a complie error.
- `[]T` is not a noreturn-like type and is treated like `struct { ptr: [*]T, len: usize }`

**Comment by andrewrk** [comment]
This proposal needs a set of test cases to act as acceptance criteria and so we have some very specific semantics to decide between.

**Comment by andrewCodeDev** [comment]
This is a small nitpick, but I'd suggest a name like "vacuous types" instead of noreturn types. Because, you can in fact return them, they just happen to be empty.

**Comment by SeanTUT** [comment]
I am currently working on a branch to tackle this. Starting out, I only aimed to address #3257, but in working on this, I found a substantial amount of buggy behavior regarding many of the types described here. I'd like to bump this discussion, since some of these semantics are undecided upon. Here are some of the choices I have made in my branch. This is early draft stuff, and I am very much open to feedback:
- `noreturn`-like types are referred to as "vacuous types"
- Vacuous types are generally treated with the same semantics as `noreturn` is today.
  - Cannot be initialized, casted to, coerced to, or stored in any way, even with `undefined`
  - An optional of an vacuous type is a OPV value which may only
  - Union fields may contain these, in which case, a switch branch on this field will go unevaluated
  - An error union may contain these as a payload, in which case, the value can only be an error
- Vacuous types are any of the following:
  - `noreturn`
  - Exhaustive enums with no fields: `enum {}`
  - Any `union` type with either no fields or all vacuous-type fields
    - This includes unions tagged by vacuous-type enums
    - Open question: should `packed`/`extern` unions be allowed to be vacuous types?
  - Vacuous error sets: `error{}`
  - Error unions where both the payload and error set are vacuous-types: `error{}!noreturn`
- Exhaustive enums with no fields must be tagged by `noreturn` (as described in #19855)
- Vacuous types are not allowed in structs, arrays, or vectors. Additionally, pointers to vacuous types are disallowed altogether.
  - I made this choice to resolve the ambiguity of what these concepts would even mean. In my opinion, it is unclear what something like array of vacuous-type values, or a pointer to an vacuous-type value would even mean.
  - It may be desirable to allow 0-length arrays of pointers to vacous types.

**Comment by rohlem** [comment]
@SeanTUT Most of what you wrote reads to me like the text in the OP, except for:

> * Vacuous types are not allowed in structs, arrays, or vectors. Additionally, pointers to vacuous types are disallowed altogether.

As per the OP,

> * structs containing any noreturn-like field
> * `[n]T` / `@Vector(n, T)` where `T` is noreturn-like and `n != 0`
> * `*T` / `[*]T` (and any qualifiers) where `T` is noreturn-like

are themselves `noreturn`-like/uninstantiable/vacuous.
That means they should be supported as `comptime` values, only instantiating them / constructing an object and accessing a storage location should result in a compile error.
(As a rule of thumb, they have 0 states. That means in sum types adding them as +0 keeps the number of states, and in product types we multiply as *0 and get another 0-state / stateless type. Neither construct would by itself a compile error.)

**Comment by SeanTUT** [comment]
> As per the OP,
> 
> > * structs containing any noreturn-like field
> > * `[n]T` / `@Vector(n, T)` where `T` is noreturn-like and `n != 0`
> > * `*T` / `[*]T` (and any qualifiers) where `T` is noreturn-like
> 
> are themselves `noreturn`-like/uninstantiable/vacuous. That means they should be supported as `comptime` values, only instantiating them / constructing an object and accessing a storage location should result in a compile error. 

Right, my implementation does differ from mlugg's original proposal. I chose this firstly for the sake of simplicity and disambiguity, but also because it aligns with the current behavior of `noreturn`. Furthermore, packed containers would not make sense with vacous types. However, I think 

Also, while I'm replying, I would like to provide my reasoning for some of my other choices:
- > pointers to vacuous types are disallowed altogether.
  - I chose this because the meaning of a pointer to a vacuous type seemed ambiguous (as I mentioned above), but I've been going back and forth on this in my implementation. Right now, I'm leaning toward making this legal.
- >`noreturn`-like types are referred to as "vacuous types"
  - As suggested by @andrewCodeDev 
  - Why not "noreturn-like?"
    - I chose not to use the term "noreturn-like" in the compiler as the word "noreturn" is typically associated with functions, even though this is not necessarily the case.
   - Why not "empty type?"
     - I chose not to use the term "empty type" because it could lead to confusion regarding container types where the word "empty" could mean something else. For example, an "empty union" (`union {}`) is an "empty-type union," but an "empty-type union" is not necessarily an "empty union."
   - Why not "zero type?"
     - I chose not to use the term "zero type" because it could lead to confusion with numeric types, nullable types, or otherwise zeroable types.

**Comment by mlugg** [comment]
A (fairly uncontroversial) subset of this behavior was implemented in https://codeberg.org/ziglang/zig/pulls/31403. I'll close this for now, though I'll probably end up revisiting these semantics more precisely at some point, in which case I'll open a new issue on Codeberg.
